packages/storage-legacy/src/LegacyAsyncStorage.ts

```diff
@@ -0,0 +1,188 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ */
+
+import {NativeModules} from 'react-native';
+import {IStorageBackend, StorageOptions} from '../../core/types';
+
+export default class LegacyAsyncStorage<VAL = any>
+  implements IStorageBackend<VAL> {
+  private readonly _asyncStorageNativeModule: any;
+
+  constructor() {
+    this._asyncStorageNativeModule =
+      NativeModules.RNC_AsyncSQLiteDBStorage ||
+      NativeModules.RNCAsyncStorage ||
+      NativeModules.PlatformLocalStorage;
+
+    if (!this._asyncStorageNativeModule) {
+      throw new Error(
+        '[@RNC/AsyncStorage]: NativeModule: AsyncStorage is null.',
+      );
+    }
+  }
+
+  static convertErrors(errs?: Array<Error> | Error) {
+    if (!errs) {
+      return null;
+    }
+    return Array.isArray(errs) ? errs.filter(e => !!e) : [errs];
+  }
+
+  async getSingle(key: string, opts?: StorageOptions): Promise<VAL | null> {
+    if (opts) {
+      // noop
+    }
+    return this._asyncStorageNativeModule.multiGet([key], function(
+      errors: Array<Error>,
+      result: any,
+    ) {
+      const value = result && result[0] && result[0][1] ? result[0][1] : null;
+      const errs = LegacyAsyncStorage.convertErrors(errors);
+      if (errs) {
+        throw errs[0];
+      } else {
+        return value;
+      }
+    });
+  }
+
+  async setSingle(
+    key: string,
+    value: VAL,
+    opts?: StorageOptions,
+  ): Promise<void> {
+    if (opts) {
+      // noop
+    }
+    this._asyncStorageNativeModule.multiSet([[key, value]], function(
```

Comment:
Note: We have to wrap every call into a Promise and return it, and resolve/reject inside that promise.